Resteasy Reactive: `ServerExceptionMapper` does not process a thrown `Exception` (Kotlin only)

### Describe the bug

If a (Kotlin-)Rest-Endpoint throws an `java.lang.Exception` then the `Exception` does not get processed by a custom `ServerExceptionMapper`. `RuntimeException` works as excepted.

```kotlin
org.acme.MyException
	at org.acme.ExceptionTestResource.throwException1Method(ExceptionTestResource.kt:33)
	at org.acme.ExceptionTestResource_Subclass.throwException1Method$$superforward1(ExceptionTestResource_Subclass.zig:113)
	at org.acme.ExceptionTestResource_Subclass$$function$$1.apply(ExceptionTestResource_Subclass$$function$$1.zig:24)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:54)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.proceed(InvocationInterceptor.java:62)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.monitor(InvocationInterceptor.java:49)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor_Bean.intercept(InvocationInterceptor_Bean.zig:521)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:41)
	at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:32)
	at org.acme.ExceptionTestResource_Subclass.throwException1Method(ExceptionTestResource_Subclass.zig:183)
	at org.acme.ExceptionTestResource_ClientProxy.throwException1Method(ExceptionTestResource_ClientProxy.zig:185)
	at org.acme.ExceptionTestResource$quarkusrestinvoker$throwException1Method_4a04d16e4f5785de322606c51c7fc425b79facd0.invoke(ExceptionTestResource$quarkusrestinvoker$throwException1Method_4a04d16e4f5785de322606c51c7fc425b79facd0.zig:33)
	at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
	at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:7)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:141)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$13.runWith(VertxCoreRecorder.java:543)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
Resulted in: io.quarkus.arc.ArcUndeclaredThrowableException: Error invoking subclass method
	at org.acme.ExceptionTestResource_Subclass.throwException1Method(ExceptionTestResource_Subclass.zig:205)
	... 12 more
```

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

1. https://code.quarkus.io/?e=resteasy-reactive-jackson&e=kotlin&extension-search=kotlin
2. Add the following code

```kotlin
package org.acme

import org.jboss.resteasy.reactive.server.ServerExceptionMapper
import javax.enterprise.context.ApplicationScoped
import javax.ws.rs.GET
import javax.ws.rs.Path
import javax.ws.rs.core.Response

open class SuperException : Exception() // Works if SuperException extends RuntimeException instead of Exception
class MyException : SuperException()

class ExceptionMappers {

    @ServerExceptionMapper
    fun superExceptionHandler(ex: SuperException): Response {
        println("SuperException processed")
        return Response.serverError().build()
    }
}

@ApplicationScoped
@Path("/ex")
class ExceptionTestResource {

    /**
     * Exception is not processed by the ServerExceptionMapper.
     */
    @GET
    @Path("/throw1")
    fun throwException1Method(): String {
        throw MyException()
    }

    /**
     * Exception is processed by the ServerExceptionMapper.
     */
    @GET
    @Path("/throw2")
    @Throws(MyException::class) // see here
    fun throwExceptionMethod(): String {
        throw MyException()
    }
}
```

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

Java 11

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

Quarkus 2.2.3.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Maven 3.8.1

### Additional information

_No response_

I've made a reproducer (see attached), including some tests.

Please note that unfortunately __the tests do not fail__. The problem can only be reproduced in __dev mode__.

Run `./mvnw quarkus:dev` call `http://localhost:8080/ex/throw1` then you get the following exception (console log):

```java
2021-09-23 16:01:39,381 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-0) HTTP Request to /ex/throw1 failed, error id: 4c3eb23b-e454-43fb-9c2c-96d44e308b1f-1: io.quarkus.arc.ArcUndeclaredThrowableException: Error invoking subclass method
	at org.acme.ExceptionTestResource_Subclass.throwException1Method(ExceptionTestResource_Subclass.zig:205)
	at org.acme.ExceptionTestResource_ClientProxy.throwException1Method(ExceptionTestResource_ClientProxy.zig:185)
	at org.acme.ExceptionTestResource$quarkusrestinvoker$throwException1Method_4a04d16e4f5785de322606c51c7fc425b79facd0.invoke(ExceptionTestResource$quarkusrestinvoker$throwException1Method_4a04d16e4f5785de322606c51c7fc425b79facd0.zig:33)
	at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
	at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:7)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:141)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$13.runWith(VertxCoreRecorder.java:543)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: org.acme.MyException
	at org.acme.ExceptionTestResource.throwException1Method(ExceptionTestResource.kt:30)
	at org.acme.ExceptionTestResource_Subclass.throwException1Method$$superforward1(ExceptionTestResource_Subclass.zig:113)
	at org.acme.ExceptionTestResource_Subclass$$function$$1.apply(ExceptionTestResource_Subclass$$function$$1.zig:24)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:54)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.proceed(InvocationInterceptor.java:62)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.monitor(InvocationInterceptor.java:49)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor_Bean.intercept(InvocationInterceptor_Bean.zig:521)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:41)
	at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:32)
	at org.acme.ExceptionTestResource_Subclass.throwException1Method(ExceptionTestResource_Subclass.zig:183)
	... 12 more

2021-09-23 16:01:39,390 ERROR [org.jbo.res.rea.com.cor.AbstractResteasyReactiveContext] (executor-thread-0) Request failed: io.quarkus.arc.ArcUndeclaredThrowableException: Error invoking subclass method
	at org.acme.ExceptionTestResource_Subclass.throwException1Method(ExceptionTestResource_Subclass.zig:205)
	at org.acme.ExceptionTestResource_ClientProxy.throwException1Method(ExceptionTestResource_ClientProxy.zig:185)
	at org.acme.ExceptionTestResource$quarkusrestinvoker$throwException1Method_4a04d16e4f5785de322606c51c7fc425b79facd0.invoke(ExceptionTestResource$quarkusrestinvoker$throwException1Method_4a04d16e4f5785de322606c51c7fc425b79facd0.zig:33)
	at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
	at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:7)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:141)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$13.runWith(VertxCoreRecorder.java:543)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: org.acme.MyException
	at org.acme.ExceptionTestResource.throwException1Method(ExceptionTestResource.kt:30)
	at org.acme.ExceptionTestResource_Subclass.throwException1Method$$superforward1(ExceptionTestResource_Subclass.zig:113)
	at org.acme.ExceptionTestResource_Subclass$$function$$1.apply(ExceptionTestResource_Subclass$$function$$1.zig:24)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:54)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.proceed(InvocationInterceptor.java:62)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.monitor(InvocationInterceptor.java:49)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor_Bean.intercept(InvocationInterceptor_Bean.zig:521)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:41)
	at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:32)
	at org.acme.ExceptionTestResource_Subclass.throwException1Method(ExceptionTestResource_Subclass.zig:183)
	... 12 more
```

[reproducer20357.zip](https://github.com/quarkusio/quarkus/files/7218182/reproducer20357.zip)

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/ResteasyReactiveScanningProcessor.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/ResteasyReactiveScanningProcessor.java
@@ -42,6 +42,7 @@ import org.jboss.resteasy.reactive.server.processor.scanning.ResteasyReactiveExc
 import org.jboss.resteasy.reactive.server.processor.scanning.ResteasyReactiveFeatureScanner;
 import org.jboss.resteasy.reactive.server.processor.scanning.ResteasyReactiveParamConverterScanner;
 
+import io.quarkus.arc.ArcUndeclaredThrowableException;
 import io.quarkus.arc.deployment.AdditionalBeanBuildItem;
 import io.quarkus.arc.deployment.GeneratedBeanBuildItem;
 import io.quarkus.arc.deployment.GeneratedBeanGizmoAdaptor;
@@ -109,6 +110,7 @@ public class ResteasyReactiveScanningProcessor {
 
         exceptions.addBlockingProblem(BlockingOperationNotAllowedException.class);
         exceptions.addBlockingProblem(BlockingNotAllowedException.class);
+        exceptions.addUnwrappedException(ArcUndeclaredThrowableException.class);
         if (capabilities.isPresent(Capability.HIBERNATE_REACTIVE)) {
             exceptions.addNonBlockingProblem(
                     new ExceptionMapping.ExceptionTypeAndMessageContainsPredicate(IllegalStateException.class, "HR000068"));

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/customexceptions/UnwrappedExceptionTest.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/customexceptions/UnwrappedExceptionTest.java
@@ -1,0 +1,88 @@
+package io.quarkus.resteasy.reactive.server.test.customexceptions;
+
+import static io.quarkus.resteasy.reactive.server.test.ExceptionUtil.removeStackTrace;
+
+import java.util.function.Supplier;
+
+import javax.ws.rs.Path;
+import javax.ws.rs.core.Response;
+
+import org.jboss.resteasy.reactive.server.ServerExceptionMapper;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.arc.ArcUndeclaredThrowableException;
+import io.quarkus.resteasy.reactive.server.test.ExceptionUtil;
+import io.quarkus.test.QuarkusUnitTest;
+import io.restassured.RestAssured;
+
+public class UnwrappedExceptionTest {
+
+    @RegisterExtension
+    static QuarkusUnitTest test = new QuarkusUnitTest()
+            .setArchiveProducer(new Supplier<>() {
+                @Override
+                public JavaArchive get() {
+                    return ShrinkWrap.create(JavaArchive.class)
+                            .addClasses(ExceptionResource.class, ExceptionMappers.class, ExceptionUtil.class);
+                }
+            });
+
+    @Test
+    public void testWrapperWithUnmappedException() {
+        RestAssured.get("/hello/wrapperOfIAE")
+                .then().statusCode(500);
+    }
+
+    @Test
+    public void testWrapperWithMappedException() {
+        RestAssured.get("/hello/wrapperOfISE")
+                .then().statusCode(999);
+    }
+
+    @Test
+    public void testUnmappedException() {
+        RestAssured.get("/hello/iae")
+                .then().statusCode(500);
+    }
+
+    @Test
+    public void testMappedException() {
+        RestAssured.get("/hello/ise")
+                .then().statusCode(999);
+    }
+
+    @Path("hello")
+    public static class ExceptionResource {
+
+        @Path("wrapperOfIAE")
+        public String wrapperOfIAE() {
+            throw removeStackTrace(new ArcUndeclaredThrowableException(removeStackTrace(new IllegalArgumentException())));
+        }
+
+        @Path("wrapperOfISE")
+        public String wrapperOfISE() {
+            throw removeStackTrace(new ArcUndeclaredThrowableException(removeStackTrace(new IllegalStateException())));
+        }
+
+        @Path("iae")
+        public String iae() {
+            throw removeStackTrace(new IllegalArgumentException());
+        }
+
+        @Path("ise")
+        public String ise() {
+            throw removeStackTrace(new IllegalStateException());
+        }
+    }
+
+    public static class ExceptionMappers {
+
+        @ServerExceptionMapper
+        Response mapISE(IllegalStateException e) {
+            return Response.status(999).build();
+        }
+    }
+}

--- a/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/ExceptionMapping.java
+++ b/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/ExceptionMapping.java
@@ -3,12 +3,15 @@ package org.jboss.resteasy.reactive.server.core;
 import io.smallrye.common.annotation.Blocking;
 import io.smallrye.common.annotation.NonBlocking;
 import java.io.IOException;
+import java.util.AbstractMap;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
+import java.util.Set;
 import java.util.function.Function;
 import java.util.function.Predicate;
 import java.util.stream.Collectors;
@@ -36,6 +39,7 @@ public class ExceptionMapping {
      */
     private final List<Predicate<Throwable>> blockingProblemPredicates = new ArrayList<>();
     private final List<Predicate<Throwable>> nonBlockingProblemPredicate = new ArrayList<>();
+    private final Set<Class<? extends Throwable>> unwrappedExceptions = new HashSet<>();
 
     @SuppressWarnings({ "unchecked", "rawtypes" })
     public void mapException(Throwable throwable, ResteasyReactiveRequestContext context) {
@@ -54,23 +58,26 @@ public class ExceptionMapping {
         }
 
         // we match superclasses only if not a WebApplicationException according to spec 3.3.4 Exceptions
-        ExceptionMapper exceptionMapper = getExceptionMapper((Class<Throwable>) klass, context);
-        if (exceptionMapper != null) {
+        Map.Entry<Throwable, ExceptionMapper<? extends Throwable>> entry = getExceptionMapper((Class<Throwable>) klass, context,
+                throwable);
+        if (entry != null) {
+            ExceptionMapper exceptionMapper = entry.getValue();
+            Throwable mappedException = entry.getKey();
             context.requireCDIRequestScope();
             if (exceptionMapper instanceof ResteasyReactiveAsyncExceptionMapper) {
-                ((ResteasyReactiveAsyncExceptionMapper) exceptionMapper).asyncResponse(throwable,
+                ((ResteasyReactiveAsyncExceptionMapper) exceptionMapper).asyncResponse(mappedException,
                         new AsyncExceptionMapperContextImpl(context));
-                logBlockingErrorIfRequired(throwable, context);
-                logNonBlockingErrorIfRequired(throwable, context);
+                logBlockingErrorIfRequired(mappedException, context);
+                logNonBlockingErrorIfRequired(mappedException, context);
                 return;
             } else if (exceptionMapper instanceof ResteasyReactiveExceptionMapper) {
-                response = ((ResteasyReactiveExceptionMapper) exceptionMapper).toResponse(throwable, context);
+                response = ((ResteasyReactiveExceptionMapper) exceptionMapper).toResponse(mappedException, context);
             } else {
-                response = exceptionMapper.toResponse(throwable);
+                response = exceptionMapper.toResponse(mappedException);
             }
             context.setResult(response);
-            logBlockingErrorIfRequired(throwable, context);
-            logNonBlockingErrorIfRequired(throwable, context);
+            logBlockingErrorIfRequired(mappedException, context);
+            logNonBlockingErrorIfRequired(mappedException, context);
             return;
         }
         if (isWebApplicationException) {
@@ -142,20 +149,26 @@ public class ExceptionMapping {
     }
 
     /**
-     * Return the proper Exception that handles {@param throwable} or {@code null}
+     * Return the proper Exception that handles {@param clazz} or {@code null}
      * if none is found.
-     * First checks if the Resource class that contained the Resource method contained class-level exception mappers
+     * First checks if the Resource class that contained the Resource method contained class-level exception mappers.
+     * {@param throwable} is optional and is used to when no mapper has been found for the original exception type, but the
+     * application
+     * has been configured to unwrap certain exceptions.
      */
-    public <T extends Throwable> ExceptionMapper<T> getExceptionMapper(Class<T> clazz, ResteasyReactiveRequestContext context) {
+    public <T extends Throwable> Map.Entry<Throwable, ExceptionMapper<? extends Throwable>> getExceptionMapper(Class<T> clazz,
+            ResteasyReactiveRequestContext context,
+            T throwable) {
         Map<Class<? extends Throwable>, ResourceExceptionMapper<? extends Throwable>> classExceptionMappers = getClassExceptionMappers(
                 context);
         if ((classExceptionMappers != null) && !classExceptionMappers.isEmpty()) {
-            ExceptionMapper<T> result = doGetExceptionMapper(clazz, classExceptionMappers);
+            Map.Entry<Throwable, ExceptionMapper<? extends Throwable>> result = doGetExceptionMapper(clazz,
+                    classExceptionMappers, throwable);
             if (result != null) {
                 return result;
             }
         }
-        return doGetExceptionMapper(clazz, mappers);
+        return doGetExceptionMapper(clazz, mappers, throwable);
     }
 
     private Map<Class<? extends Throwable>, ResourceExceptionMapper<? extends Throwable>> getClassExceptionMappers(
@@ -166,19 +179,27 @@ public class ExceptionMapping {
         return context.getTarget() != null ? context.getTarget().getClassExceptionMappers() : null;
     }
 
-    @SuppressWarnings("unchecked")
-    private <T extends Throwable> ExceptionMapper<T> doGetExceptionMapper(Class<T> clazz,
-            Map<Class<? extends Throwable>, ResourceExceptionMapper<? extends Throwable>> mappers) {
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    private <T extends Throwable> Map.Entry<Throwable, ExceptionMapper<? extends Throwable>> doGetExceptionMapper(
+            Class<T> clazz,
+            Map<Class<? extends Throwable>, ResourceExceptionMapper<? extends Throwable>> mappers,
+            Throwable throwable) {
         Class<?> klass = clazz;
         do {
             ResourceExceptionMapper<? extends Throwable> mapper = mappers.get(klass);
             if (mapper != null) {
-                return (ExceptionMapper<T>) mapper.getFactory()
-                        .createInstance().getInstance();
+                return new AbstractMap.SimpleEntry(throwable, mapper.getFactory()
+                        .createInstance().getInstance());
             }
             klass = klass.getSuperclass();
         } while (klass != null);
 
+        if ((throwable != null) && unwrappedExceptions.contains(clazz)) {
+            Throwable cause = throwable.getCause();
+            if (cause != null) {
+                return doGetExceptionMapper(cause.getClass(), mappers, cause);
+            }
+        }
         return null;
     }
 
@@ -196,6 +217,10 @@ public class ExceptionMapping {
 
     public void addNonBlockingProblem(Predicate<Throwable> predicate) {
         nonBlockingProblemPredicate.add(predicate);
+    }
+
+    public void addUnwrappedException(Class<? extends Throwable> clazz) {
+        unwrappedExceptions.add(clazz);
     }
 
     public <T extends Throwable> void addExceptionMapper(Class<T> exceptionClass, ResourceExceptionMapper<T> mapper) {
@@ -219,6 +244,10 @@ public class ExceptionMapping {
 
     public List<Predicate<Throwable>> getNonBlockingProblemPredicate() {
         return nonBlockingProblemPredicate;
+    }
+
+    public Set<Class<? extends Throwable>> getUnwrappedExceptions() {
+        return unwrappedExceptions;
     }
 
     public void initializeDefaultFactories(Function<String, BeanFactory<?>> factoryCreator) {

--- a/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/jaxrs/ProvidersImpl.java
+++ b/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/jaxrs/ProvidersImpl.java
@@ -3,6 +3,7 @@ package org.jboss.resteasy.reactive.server.jaxrs;
 import java.lang.annotation.Annotation;
 import java.lang.reflect.Type;
 import java.util.List;
+import java.util.Map;
 import javax.ws.rs.RuntimeType;
 import javax.ws.rs.core.MediaType;
 import javax.ws.rs.ext.ContextResolver;
@@ -44,9 +45,15 @@ public class ProvidersImpl implements Providers {
         return null;
     }
 
+    @SuppressWarnings("unchecked")
     @Override
     public <T extends Throwable> ExceptionMapper<T> getExceptionMapper(Class<T> type) {
-        return deployment.getExceptionMapping().getExceptionMapper(type, null);
+        Map.Entry<Throwable, ExceptionMapper<? extends Throwable>> entry = deployment.getExceptionMapping()
+                .getExceptionMapper(type, null, null);
+        if (entry != null) {
+            return (ExceptionMapper<T>) entry.getValue();
+        }
+        return null;
     }
 
     @Override

